runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/ActiveWorkRefresher.java

```diff
@@ -103,5 +116,35 @@ private void invalidateStuckCommits() {
     }
   }
 
-  protected abstract void refreshActiveWork();
+  private void refreshActiveWork() {
+    Instant refreshDeadline = clock.get().minus(Duration.millis(activeWorkRefreshPeriodMillis));
+
+    Map<HeartbeatSender, Heartbeat> fannedOutHeartbeatRequests = new HashMap<>();
+
+    for (ComputationState computationState : computations.get()) {
+      String computationId = computationState.getComputationId();
+
+      // Get heartbeat requests for computation's current active work, aggregated by GetDataStream
+      // to correctly fan-out the heartbeat requests.
+      Table<HeartbeatSender, RefreshableWork, HeartbeatRequest> heartbeats =
+          HeartbeatRequests.getRefreshableKeyHeartbeats(
```

Comment:
this seems to be the only use of this and it doesn't really use the table, just iterates over the cells.  Can this method instead be changed to add to the heartbeatsender -> heartbeats builder? (see other comment)

Seems like we're giong to have a lot of unneeded allocations to build the rows and tables instead of just building into what we want in the end.

Reply:
done